stdlib/public/core/StringHashable.swift

```diff
@@ -12,181 +12,129 @@
 
 import SwiftShims
 
-#if _runtime(_ObjC)
-@_inlineable // FIXME(sil-serialize-all)
-@_versioned // FIXME(sil-serialize-all)
-@_silgen_name("swift_stdlib_NSStringHashValue")
-internal func _stdlib_NSStringHashValue(
-  _ str: AnyObject, isASCII: Bool) -> Int
-
-@_inlineable // FIXME(sil-serialize-all)
-@_versioned // FIXME(sil-serialize-all)
-@_silgen_name("swift_stdlib_NSStringHashValuePointer")
-internal func _stdlib_NSStringHashValuePointer(
-  _ str: OpaquePointer, isASCII: Bool) -> Int
-
-@_inlineable // FIXME(sil-serialize-all)
-@_versioned // FIXME(sil-serialize-all)
-@_silgen_name("swift_stdlib_CFStringHashCString")
-internal func _stdlib_CFStringHashCString(
-  _ str: OpaquePointer, _ len: Int) -> Int
-#endif
-
-extension Unicode {
-  // FIXME: cannot be marked @_versioned. See <rdar://problem/34438258>
-  // @_inlineable // FIXME(sil-serialize-all)
-  // @_versioned // FIXME(sil-serialize-all)
-  internal static func hashASCII(
-    _ string: UnsafeBufferPointer<UInt8>,
-    into hasher: inout _Hasher
-  ) {
-    let collationTable = _swift_stdlib_unicode_getASCIICollationTable()
-    for c in string {
-      _precondition(c <= 127)
-      let element = collationTable[Int(c)]
-      // Ignore zero valued collation elements. They don't participate in the
-      // ordering relation.
-      if element != 0 {
-        hasher.append(Int(truncatingIfNeeded: element))
+func _emptyASCIIHashBuffer() -> _UIntBuffer<UInt64, UInt8> {
+  var buffer = _UIntBuffer<UInt64, UInt8>()
+  // null terminated strings need to hash differently than non-null terminated
+  // ones. A partially filled ascii buffer should have 1s in the leftover space
+  buffer._storage = UInt64.max
+  return buffer
+}
+
+extension _UnmanagedString where CodeUnit == UInt8 {
+  // NOT @_versioned
+  @effects(releasenone)
+  internal func hashASCII(into hasher: inout _Hasher) {
+    var buffer = _emptyASCIIHashBuffer()
+    for c in self {
+      if buffer.count < buffer.capacity {
+        buffer.append(UInt8(truncatingIfNeeded: c))
+      }
+
+      if buffer.count == buffer.capacity {
+        hasher.append(buffer._storage)
+        buffer = _emptyASCIIHashBuffer()
```

Comment:
Given that this is such a frequently duplicated block of code, why not add it to an ASCIIHasher type?